### PR TITLE
Migrating build tooling to Webpack 5

### PR DESCRIPTION
**Pull request by benjdlambert**
Signed-off-by: blam <ben@blam.sh>

https://github.com/webpack/webpack/pull/8460/commits/a68426e9255edcce7822480b78416837617ab065

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/lib/bundler/config.ts [hunk 1]
```diff
@@ -110,6 +110,12 @@ export async function createConfig(
     );
   }
 
+  plugins.push(
+    new ProvidePlugin({
+      process: 'process/browser',
```
Comment (Rugvip) [inline packages/cli/src/lib/bundler/config.ts:115]: What's the story behind this one, maybe a comment? :>

**Review by Rugvip — COMMENTED** [review]
Did some profiling of an app build, not looking great 😅 

<summary>
Webpack v4
<details>

```
<w> [webpack.Progress] 27415ms building
<i> [webpack.Progress] 70ms finish module graph
<i> [webpack.Progress] 157ms dependencies optimization
<i> [webpack.Progress] 97ms chunk graph
<i> [webpack.Progress] 27ms basic chunk optimization
<i> [webpack.Progress] 115ms advanced chunk optimization
<i> [webpack.Progress] 770ms chunk modules optimization
<i> [webpack.Progress] 27ms module order optimization
<i> [webpack.Progress] 27ms module id optimization
<i> [webpack.Progress] 25ms record modules
<i> [webpack.Progress] 197ms hashing
<w> [webpack.Progress] 1618ms chunk assets processing
<i> [webpack.Progress] 17ms additional chunk assets processing
<w> [webpack.Progress] 3643ms chunk asset optimization
<w> [webpack.Progress] 15685ms after chunk asset optimization
<i> [webpack.Progress] 293ms emitting
<i> [webpack.Progress] 74ms after emitting
  354.86 KB  dist/static/vendor.7d25bf33.chunk.js
  260.6 KB   dist/static/module-highlight.js.3d0df306.js
  226.59 KB  dist/static/main.c2af6e31.chunk.js
  169.74 KB  dist/static/module-refractor.adbe310d.js
  152.07 KB  dist/static/module-swagger-ui-react.030bbb48.js
  149.11 KB  dist/static/2.456f7b9b.chunk.js
  118.52 KB  dist/static/module-material-ui.86cac652.js
  112.8 KB   dist/static/module-jspdf.541556af.js
  70.15 KB   dist/static/module-moment.5e6d641f.js
  63.65 KB   dist/static/module-lodash.5c596726.js
  61.15 KB   dist/static/1.e51e842b.chunk.js
  53.83 KB   dist/static/module-recharts.9ff14e4d.js
  50.23 KB   dist/static/39.880da7f6.chunk.js
  47.79 KB   dist/static/module-ajv.efe54267.js
  47.54 KB   dist/static/38.03900e94.chunk.js
  44.48 KB   dist/static/module-asyncapi.ba886d73.js
  41.79 KB   dist/static/module-roadiehq.19597a66.js
  40.49 KB   dist/static/module-luxon.d3a99300.js
  39.68 KB   dist/static/54.b60f3a86.chunk.js
  37.04 KB   dist/static/module-graphql.6fb6a9f7.js
  35.77 KB   dist/static/module-hot-loader.0c60ebc4.js
  28.74 KB   dist/static/module-esprima.3984480a.js
  28.43 KB   dist/static/module-react-beautiful-dnd.4e6fe353.js
  28.25 KB   dist/static/module-yaml.c9b27cb5.js
  27.68 KB   dist/static/module-material-table.f6197a6b.js
  21.54 KB   dist/static/swagger-ui-react.34.fe7e0c0b.css
  20.82 KB   dist/static/56.ea4e8193.chunk.js
  20.72 KB   dist/static/module-js-yaml.6874c9f5.js
  20.06 KB   dist/static/module-react-syntax-highlighter.f5fe07d2.js
  18.59 KB   dist/static/module-autolinker.b1b9bc50.js
  16.71 KB   dist/static/module-sanitize-html.15b96def.js
  15.31 KB   dist/static/module-immutable.315648d4.js
  14.66 KB   dist/static/module-octokit.d70ad18b.js
  13.51 KB   dist/static/module-remarkable.e80120b6.js
  13.3 KB    dist/static/module-yaml-ast-parser.502b25d1.js
  11.82 KB   dist/static/module-date-fns.d9a4ffb9.js
  11.78 KB   dist/static/module-micromark.fbb17f27.js
  11.03 KB   dist/static/module-htmlparser2.202d391e.js
  10.23 KB   dist/static/module-json-to-ast.3d89c5cf.js
  9.41 KB    dist/static/52.f4a603ad.chunk.js
  9.27 KB    dist/static/module-lodash-es.4e5db5ee.js
  7.67 KB    dist/static/module-source-map.32a506ee.js
  7.59 KB    dist/static/vendor.35.8c496d5a.css
  4.71 KB    dist/static/asyncapi.4.a98603d5.css
  4.4 KB     dist/static/55.5f909df8.chunk.js
  2 KB       dist/static/40.d06c6c05.chunk.js
  1.68 KB    dist/static/49.68dc62d5.chunk.js
  1.36 KB    dist/static/runtime.d917d5b5.js
  1.22 KB    dist/static/51.d1cd3859.chunk.js
  1.04 KB    dist/static/41.1a97c820.chunk.js
  959 B      dist/static/53.5c8bea4d.chunk.js
  742 B      dist/static/44.cd0901f3.chunk.js
  695 B      dist/static/47.4c8f4541.chunk.js
  683 B      dist/static/46.e1d21803.chunk.js
  660 B      dist/static/48.c025426f.chunk.js
  289 B      dist/static/50.411388d1.chunk.js
  246 B      dist/static/0.996f2136.chunk.js
  244 B      dist/static/45.9eeb8235.chunk.js
  213 B      dist/static/42.73b9c56e.chunk.js
  193 B      dist/static/43.5d12f749.chunk.js
✨  Done in 55.73s.
```

</details>
</summary>


<summary>
Webpack v5
<details>

```
<i> [webpack.Progress]  | 687 ms setup > initialize
<i> [webpack.Progress]  |  | 13 ms compilation > EntryPlugin
<i> [webpack.Progress]  | 40 ms setup > compilation
<i> [webpack.Progress] 740 ms setup
<i> [webpack.Progress]  | 146 ms building > entries dependencies modules
<i> [webpack.Progress]  | 79 ms building > import loader ../../node_modules/@sucrase/webpack-loader/dist/index.js
<w> [webpack.Progress]  | 1528 ms building > entries dependencies modules
<w> [webpack.Progress]  | 1853 ms building > import loader ../../node_modules/@svgr/webpack/lib/index.js
<w> [webpack.Progress]  | 5391 ms building > entries dependencies modules
<i> [webpack.Progress]  | 158 ms building > import loader ../../node_modules/css-loader/dist/cjs.js
<i> [webpack.Progress]  | 489 ms building > entries dependencies modules
<i> [webpack.Progress]  | 35 ms building > import loader ../../node_modules/url-loader/dist/cjs.js
<e> [webpack.Progress]  | 12884 ms building > entries dependencies modules
<e> [webpack.Progress] 22567 ms building
<i> [webpack.Progress]  |  | 305 ms finish module graph > FlagDependencyExportsPlugin
<i> [webpack.Progress]  | 409 ms sealing > finish module graph
<i> [webpack.Progress]  |  | 84 ms dependencies optimization > SideEffectsFlagPlugin
<i> [webpack.Progress]  |  | 323 ms dependencies optimization > FlagDependencyUsagePlugin
<i> [webpack.Progress]  | 407 ms sealing > dependencies optimization
<i> [webpack.Progress]  | 273 ms sealing > chunk graph
<i> [webpack.Progress]  |  | 178 ms chunk optimization > SplitChunksPlugin
<i> [webpack.Progress]  | 184 ms sealing > chunk optimization
<i> [webpack.Progress]  |  | 503 ms chunk modules optimization > ModuleConcatenationPlugin
<i> [webpack.Progress]  | 503 ms sealing > chunk modules optimization
<i> [webpack.Progress]  |  | 209 ms module ids > DeterministicModuleIdsPlugin
<i> [webpack.Progress]  | 209 ms sealing > module ids
<i> [webpack.Progress]  |  | 18 ms chunk ids > DeterministicChunkIdsPlugin
<i> [webpack.Progress]  | 18 ms sealing > chunk ids
<i> [webpack.Progress]  |  | 15 ms chunk id optimization > FlagIncludedChunksPlugin
<i> [webpack.Progress]  | 15 ms sealing > chunk id optimization
<i> [webpack.Progress]  |  | 14 ms record modules > RecordIdsPlugin
<i> [webpack.Progress]  | 15 ms sealing > record modules
<i> [webpack.Progress]  | 87 ms sealing > record chunks
<i> [webpack.Progress]  | 444 ms sealing > module hashing
<w> [webpack.Progress]  | 2904 ms sealing > code generation
<i> [webpack.Progress]  | 24 ms sealing > runtime requirements
<i> [webpack.Progress]  | 71 ms sealing > hashing
<i> [webpack.Progress]  | 64 ms sealing > chunk assets processing
92% sealing asset processing PersistentChildCompilerSingletonPlugin(node:14212) [DEP_WEBPACK_TEMPLATE_PATH_PLUGIN_REPLACE_PATH_VARIABLES_HASH] DeprecationWarning: [hash] is now [fullhash] (also consider using [chunkhash] or [contenthash], see documentation for details)
<e> [webpack.Progress]  |  | 77224 ms asset processing > TerserPlugin
<w> [webpack.Progress]  |  |  | 2912 ms SourceMapDevToolPlugin > static/main.c366802a.js
<i> [webpack.Progress]  |  |  | 18 ms SourceMapDevToolPlugin > static/runtime.c366802a.js
<i> [webpack.Progress]  |  |  | 227 ms SourceMapDevToolPlugin > static/1763.fcbb6c6c.chunk.js
<i> [webpack.Progress]  |  |  | 398 ms SourceMapDevToolPlugin > static/3285.d37f803c.chunk.js
<w> [webpack.Progress]  |  |  | 1120 ms SourceMapDevToolPlugin > static/3714.90609bbc.chunk.js
<i> [webpack.Progress]  |  |  | 77 ms SourceMapDevToolPlugin > static/4902.cdde54b1.chunk.js
<i> [webpack.Progress]  |  |  | 421 ms SourceMapDevToolPlugin > static/4968.683d5795.chunk.js
<i> [webpack.Progress]  |  |  | 11 ms SourceMapDevToolPlugin > static/3206.c5fb5001.chunk.js
<i> [webpack.Progress]  |  |  | 865 ms SourceMapDevToolPlugin > static/module-material-ui.3d96af86.js
<i> [webpack.Progress]  |  |  | 547 ms SourceMapDevToolPlugin > static/module-lodash.26f7b94a.js
<i> [webpack.Progress]  |  |  | 340 ms SourceMapDevToolPlugin > static/module-refractor.009ac6c0.js
<i> [webpack.Progress]  |  |  | 539 ms SourceMapDevToolPlugin > static/module-highlight.js.737987f4.js
<i> [webpack.Progress]  |  |  | 81 ms SourceMapDevToolPlugin > static/module-lodash-es.2a0858a5.js
<i> [webpack.Progress]  |  |  | 318 ms SourceMapDevToolPlugin > static/module-asyncapi.faf6f75f.js
<i> [webpack.Progress]  |  |  | 587 ms SourceMapDevToolPlugin > static/module-recharts.7ce11c9c.js
<i> [webpack.Progress]  |  |  | 639 ms SourceMapDevToolPlugin > static/module-moment.47bec17a.js
<i> [webpack.Progress]  |  |  | 218 ms SourceMapDevToolPlugin > static/module-react-syntax-highlighter.090e012f.js
<i> [webpack.Progress]  |  |  | 325 ms SourceMapDevToolPlugin > static/module-ajv.945f6c97.js
<i> [webpack.Progress]  |  |  | 274 ms SourceMapDevToolPlugin > static/module-graphql.9d76da68.js
<i> [webpack.Progress]  |  |  | 123 ms SourceMapDevToolPlugin > static/module-date-fns.a371b4ea.js
<i> [webpack.Progress]  |  |  | 87 ms SourceMapDevToolPlugin > static/module-micromark.1b3cf1b7.js
<i> [webpack.Progress]  |  |  | 210 ms SourceMapDevToolPlugin > static/module-roadiehq.dec1ddc9.js
<i> [webpack.Progress]  |  |  | 142 ms SourceMapDevToolPlugin > static/module-js-yaml.2ce1ba33.js
<i> [webpack.Progress]  |  |  | 82 ms SourceMapDevToolPlugin > static/module-yaml-ast-parser.dc49c36b.js
<i> [webpack.Progress]  |  |  | 89 ms SourceMapDevToolPlugin > static/module-sanitize-html.f5051a31.js
<i> [webpack.Progress]  |  |  | 63 ms SourceMapDevToolPlugin > static/module-autolinker.5856b746.js
<i> [webpack.Progress]  |  |  | 209 ms SourceMapDevToolPlugin > static/module-material-table.42c8e9b5.js
<i> [webpack.Progress]  |  |  | 39 ms SourceMapDevToolPlugin > static/swagger-ui-react.78209f32.css
<i> [webpack.Progress]  |  |  | 930 ms SourceMapDevToolPlugin > static/module-swagger-ui-react.f180c9b4.js
<i> [webpack.Progress]  |  |  | 185 ms SourceMapDevToolPlugin > static/module-react-beautiful-dnd.d914ddfd.js
<i> [webpack.Progress]  |  |  | 68 ms SourceMapDevToolPlugin > static/module-htmlparser2.eaa53f99.js
<i> [webpack.Progress]  |  |  | 58 ms SourceMapDevToolPlugin > static/module-octokit.40657b07.js
<i> [webpack.Progress]  |  |  | 42 ms SourceMapDevToolPlugin > static/module-source-map.55032cfb.js
<i> [webpack.Progress]  |  |  | 166 ms SourceMapDevToolPlugin > static/module-yaml.68e40086.js
<i> [webpack.Progress]  |  |  | 290 ms SourceMapDevToolPlugin > static/module-luxon.09c03c24.js
<i> [webpack.Progress]  |  |  | 178 ms SourceMapDevToolPlugin > static/module-hot-loader.2a75a9b0.js
<i> [webpack.Progress]  |  |  | 91 ms SourceMapDevToolPlugin > static/module-remarkable.d965d93a.js
<i> [webpack.Progress]  |  |  | 631 ms SourceMapDevToolPlugin > static/module-jspdf.cc8eceb3.js
<i> [webpack.Progress]  |  |  | 95 ms SourceMapDevToolPlugin > static/module-json-to-ast.b3cc70df.js
<i> [webpack.Progress]  |  |  | 106 ms SourceMapDevToolPlugin > static/module-immutable.91b12039.js
<i> [webpack.Progress]  |  |  | 164 ms SourceMapDevToolPlugin > static/module-esprima.e514074d.js
<i> [webpack.Progress]  |  |  | 26 ms SourceMapDevToolPlugin > static/vendor.3b84b402.css
<w> [webpack.Progress]  |  |  | 2865 ms SourceMapDevToolPlugin > static/vendor.c366802a.js
<i> [webpack.Progress]  |  |  | 328 ms SourceMapDevToolPlugin > static/9853.47c6335c.chunk.js
<i> [webpack.Progress]  |  |  | 141 ms SourceMapDevToolPlugin > static/9339.f467738a.chunk.js
<i> [webpack.Progress]  |  |  | 26 ms SourceMapDevToolPlugin > static/main.c366802a.js
<i> [webpack.Progress]  |  |  | 12 ms SourceMapDevToolPlugin > static/3714.90609bbc.chunk.js
<i> [webpack.Progress]  |  |  | 11 ms SourceMapDevToolPlugin > static/4968.683d5795.chunk.js
<i> [webpack.Progress]  |  |  | 16 ms SourceMapDevToolPlugin > static/module-material-ui.3d96af86.js
<i> [webpack.Progress]  |  |  | 11 ms SourceMapDevToolPlugin > static/module-recharts.7ce11c9c.js
<i> [webpack.Progress]  |  |  | 11 ms SourceMapDevToolPlugin > static/module-date-fns.a371b4ea.js
<i> [webpack.Progress]  |  |  | 33 ms SourceMapDevToolPlugin > static/vendor.c366802a.js
<e> [webpack.Progress]  |  | 17611 ms asset processing > SourceMapDevToolPlugin
<i> [webpack.Progress]  |  | 134 ms asset processing > HtmlWebpackPlugin
<i> [webpack.Progress]  |  | 135 ms asset processing > RealContentHashPlugin
<e> [webpack.Progress]  | 95106 ms sealing > asset processing
<i> [webpack.Progress]  | 16 ms sealing > after asset optimization
<e> [webpack.Progress] 100764 ms sealing
<i> [webpack.Progress]  | 163 ms emitting > emit
<i> [webpack.Progress] 167 ms emitting
  339.23 KB  dist/static/vendor.c366802a.js
  260.09 KB  dist/static/module-highlight.js.737987f4.js
  215.48 KB  dist/static/main.c366802a.js
  170.77 KB  dist/static/module-refractor.009ac6c0.js
  153.3 KB   dist/static/module-swagger-ui-react.f180c9b4.js
  145.37 KB  dist/static/3714.90609bbc.chunk.js
  114.55 KB  dist/static/module-material-ui.3d96af86.js
  106.3 KB   dist/static/module-jspdf.cc8eceb3.js
  70.41 KB   dist/static/module-moment.47bec17a.js
  64.61 KB   dist/static/module-lodash.26f7b94a.js
  61.16 KB   dist/static/4968.683d5795.chunk.js
  53.94 KB   dist/static/module-recharts.7ce11c9c.js
  50.59 KB   dist/static/3285.d37f803c.chunk.js
  48.69 KB   dist/static/9853.47c6335c.chunk.js
  46.75 KB   dist/static/module-ajv.945f6c97.js
  43.43 KB   dist/static/module-asyncapi.faf6f75f.js
  43.08 KB   dist/static/module-roadiehq.dec1ddc9.js
  40.32 KB   dist/static/module-luxon.09c03c24.js
  39.39 KB   dist/static/1763.fcbb6c6c.chunk.js
  37.11 KB   dist/static/module-graphql.9d76da68.js
  35.7 KB    dist/static/module-hot-loader.2a75a9b0.js
  28.43 KB   dist/static/module-esprima.e514074d.js
  28.19 KB   dist/static/module-react-beautiful-dnd.d914ddfd.js
  27.94 KB   dist/static/module-material-table.42c8e9b5.js
  26.89 KB   dist/static/module-yaml.68e40086.js
  22.4 KB    dist/static/module-js-yaml.2ce1ba33.js
  21.53 KB   dist/static/swagger-ui-react.51e2578a.css
  20.72 KB   dist/static/9339.f467738a.chunk.js
  19.34 KB   dist/static/module-react-syntax-highlighter.090e012f.js
  18.58 KB   dist/static/module-autolinker.5856b746.js
  16.77 KB   dist/static/module-sanitize-html.f5051a31.js
  15.31 KB   dist/static/module-immutable.91b12039.js
  13.48 KB   dist/static/module-remarkable.d965d93a.js
  13.38 KB   dist/static/module-yaml-ast-parser.dc49c36b.js
  12.33 KB   dist/static/module-octokit.40657b07.js
  12.13 KB   dist/static/module-date-fns.a371b4ea.js
  11.83 KB   dist/static/vendor.0ba22d61.css
  11.83 KB   dist/static/module-micromark.1b3cf1b7.js
  11.16 KB   dist/static/module-htmlparser2.eaa53f99.js
  10.22 KB   dist/static/module-json-to-ast.b3cc70df.js
  9.5 KB     dist/static/4902.cdde54b1.chunk.js
  9.24 KB    dist/static/module-lodash-es.2a0858a5.js
  7.7 KB     dist/static/module-source-map.55032cfb.js
  4.42 KB    dist/static/3206.c5fb5001.chunk.js
  2.03 KB    dist/static/runtime.c366802a.js
  1.99 KB    dist/static/1959.166b57b6.chunk.js
  1.71 KB    dist/static/1710.9070f9ac.chunk.js
  1.25 KB    dist/static/7422.6f54b7f4.chunk.js
  1.03 KB    dist/static/4960.96581098.chunk.js
  966 B      dist/static/8090.69e52d47.chunk.js
  759 B      dist/static/4285.39efac1a.chunk.js
  707 B      dist/static/8792.b80f1b6a.chunk.js
  695 B      dist/static/4022.82e9f6f8.chunk.js
  671 B      dist/static/390.f826e3e4.chunk.js
  298 B      dist/static/3440.f9c6f2ff.chunk.js
  243 B      dist/static/6650.9cfbc8df.chunk.js
  210 B      dist/static/9929.e0821d33.chunk.js
  179 B      dist/static/6609.60ea0b35.chunk.js
  151 B      dist/static/2687.b78f0244.chunk.js
✨  Done in 131.39s.
```

</details>
</summary>

v5 produces a larger result in more than twice the time ._.

At least it's clear where the performance hit is, so we can look into options for the terser plugin :>

**Comment by Rugvip** [comment]
Needs a changeset btw :D

File: packages/cli/src/lib/bundler/config.ts (hunk 1 above)
Comment (benjdlambert) [reply packages/cli/src/lib/bundler/config.ts:115]: Because `process` is a global not like a `import process from 'process'` polyfill. Will add a comment.

**Comment by freben** [comment]
oops, conflict now

**Comment by Rugvip** [comment]
Note for us that we're pausing this until we can get hot module reloading performance sorted, possibly waiting for react-scripts to migrate to webpack v5

**Comment by benjdlambert** [comment]
Looks like we can pick this up again soon with the versions listed in here:
https://github.com/facebook/create-react-app/issues/9994

**Comment by benjdlambert** [comment]
Looks like maybe we can take advantage of this:
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50773#issuecomment-791257620

Or wait for this issue in general
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50773

**Review by Rugvip — APPROVED** [review]

**Pull request merged**